ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,884 @@
+    }
+    checkSanity(details);
+    compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, time, details);
+    resumeCompaction();
+  }
+
+  /**
+   * Resumes compaction from where it was left off.
+   * @throws IllegalStateException if the compactor has not been initialized or if there is no compaction to resume.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void resumeCompaction() throws StoreException {
+    if (srcIndex == null) {
+      throw new IllegalStateException("Compactor has not been initialized");
+    } else if (compactionLog == null) {
+      throw new IllegalStateException("There is no compaction to resume");
+    }
+    runningLatch = new CountDownLatch(1);
+    /*
+    A single compaction job could be performed across multiple compaction cycles (if there aren't enough swap spaces to
+    complete the job in one cycle). Therefore, we loop and perform PREPARE, COPY, COMMIT and CLEANUP until the
+    CompactionLog reports the job as DONE.
+    Short description of each of the phases:
+    1. PREPARE - Initial state of a compaction cycle. Signifies that the cycle has just started.
+    2. COPY - Valid data from the segments under compaction is copied over to the swap spaces. If all the
+    data cannot be copied, it auto splits the compaction job into two cycles.
+    3. COMMIT - Adds the new log segments to the application log and atomically changes the set of index segments
+    4. CLEANUP - Cleans up the old log segments and their index files
+     */
+
+    try {
+      while (isActive && !compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+        CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+        switch (phase) {
+          case PREPARE:
+            compactionLog.markCopyStart();
+            // fall through to COPY
+          case COPY:
+            copy();
+            if (isActive) {
+              compactionLog.markCommitStart();
+              commit(false);
+              compactionLog.markCleanupStart();
+              waitTillRefCountZero();
+              cleanup(false);
+              compactionLog.markCycleComplete();
+            }
+            break;
+          default:
+            throw new IllegalStateException("Illegal compaction phase: " + phase);
+        }
+      }
+      endCompaction();
+    } catch (InterruptedException | IOException e) {
+      throw new StoreException("Exception during compaction", e, StoreErrorCodes.Unknown_Error);
+    } finally {
+      runningLatch.countDown();
+    }
+  }
+
+  /**
+   * Checks the sanity of the provided {@code details}
+   * @param details the {@link CompactionDetails} to use for compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log or if one or more offsets
+   * in the segments to compact are in the journal.
+   */
+  private void checkSanity(CompactionDetails details) {
```

Comment:
Any failure of compaction will require manual intervention to discover the root of the problem. This can only happen due to some software bug in the `CompactionManager` for e.g.